Aggs: Scripted metric allow list

alex-spies opened the pull request:
This introduces new cluster settings that enable allowing only a certain set of scripts in scripted metrics aggregations. 

By default, we permit all scripted metrics (and thus this is no breaking change).

Example:
```
PUT _cluster/settings
{
    "persistent": {
        "search.aggs.only_allowed_metric_scripts": true,
        "search.aggs.allowed_inline_metric_scripts": [
            "foo",
            "bar"
        ],
        "search.aggs.allowed_stored_metric_scripts": [
            "some-id",
            "some-other-id"
        ]
    }
}

POST /testidx/_search
{
    "aggs": {
        "sum_everything": {
            "scripted_metric": {
                "init_script": "state.transactions = []",
                "map_script": "state.transactions.add(doc.some_field.value)",
                "combine_script": "long sum = 0; for (t in state.transactions) { sum += t } return sum",
                "reduce_script": "long sum = 0; for (a in states) { sum += a } return sum"
            }
        }
    }
}
```

This leads to an IAE:

```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "illegal_argument_exception",
        "reason" : "[init_script] contains not allowed script: [sum_everything]"
      }
    ],
    "type" : "search_phase_execution_exception",
    "reason" : "all shards failed",
    "phase" : "query",
    "grouped" : true,
    "failed_shards" : [
      {
        "shard" : 0,
        "index" : "testidx",
        "node" : "jmkeNBHVRfi_8GpllAwbDw",
        "reason" : {
          "type" : "illegal_argument_exception",
          "reason" : "[init_script] contains not allowed script: [sum_everything]"
        }
      }
    ],
    "caused_by" : {
      "type" : "illegal_argument_exception",
      "reason" : "[init_script] contains not allowed script: [sum_everything]",
      "caused_by" : {
        "type" : "illegal_argument_exception",
        "reason" : "[init_script] contains not allowed script: [sum_everything]"
      }
    }
  },
  "status" : 400
}
```

The same happens when we use stored scripts:

```
POST /testidx/_search
{
    "aggs": {
        "sum_everything": {
            "scripted_metric": {
                "init_script": {"id": "init-script"},
                "map_script": {"id": "map-script"},
                "combine_script": {"id": "combine-script"},
                "reduce_script": {"id": "reduce-script"}
            }
        }
    }
}
```

Setting `search.aggs.only_allowed_metric_scripts: true` but leaving the allow lists empty prevents any scripted metric aggs from running.

alex-spies committed:
Add setting

alex-spies committed:
Move setting to SearchModule

alex-spies committed:
Turn setting into actual validation

nik9000 commented on server/src/main/java/org/elasticsearch/search/SearchModule.java:
```diff
@@ -291,6 +291,20 @@ public class SearchModule {
+        // TODO: remove, only for testing
+        Setting.Property.Dynamic
+    // TODO: Needed?
+    // Setting.Property.Final
```
I think we don't want this to be final - we want to be able to change it on the fly at least for testing.

nik9000 commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/ScriptedMetricAggregationBuilder.java:
```diff
@@ -39,21 +40,56 @@ public class ScriptedMetricAggregationBuilder extends AbstractAggregationBuilder
+                throw new IllegalArgumentException("[" + scriptName + "] contains not allowed script: [" + aggName + "]");
+            }
+        }
+    }
```
I think i'd prefer to put the allowlist in `ScriptedMetricAggregationBuilder#doBuild`. There's an `AggregationContext` which you can wire the validation logic into. When we build `ProductionAggregationContext` we already have a copy of `ClusterSettings` which I *think* is where we'd grab the list.

I feel like parsing is too early. It's not actually too early because we don't have any way around the public API, but I still think of our REST layer as a wrapper around the "real" ES so I don't want to put the security bits in it.

alex-spies replied:
Thanks! You're right, this is better.

alex-spies committed:
Remove some todos

alex-spies committed:
Revert "Turn setting into actual validation"

This reverts commit 73f8c6bc5c920eca3c33ae7c058dce9a6488a786.

alex-spies committed:
Move validation into doBuild

alex-spies committed:
Add yaml tests

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into scripted_metric_allow_list

alex-spies committed:
Fix bwc in yaml tests

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into scripted_metric_allow_list

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into scripted_metric_allow_list

alex-spies committed:
Make setting specific to inline scripts

alex-spies committed:
Stored scripts, disallow all, tests

alex-spies commented:
Update: introduced two separate settings for stored and inline metric scripts. Also added a overall switch to enable the allow list behavior at all; this is important to enable admins to disable stored scripts in the scripted metrics aggregation altogether.

Allowing even a finite list of stored scripts introduces the risk that
- we permit a stored script with id `foo`,
- a users accidentally has permissions for `PUT _scripts/foo`,
- the user replaces the script `foo` with something malicious/memory hungry, and
- this blows up the cluster(s).

alex-spies committed:
Make CoreWithMappedRuntimeFieldsIT happy

This one didn't know about the put_script api.

alex-spies committed:
Make (Rcs)CcsCommonYamlTestSuiteIT happy as well

alex-spies committed:
Update docs/changelog/109444.yaml

elasticsearchmachine commented:
Hi @alex-spies, I've created a changelog YAML for you.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into scripted_metric_allow_list
